src/stream/src/common/table/state_table.rs

```diff
@@ -776,7 +862,59 @@ where
         self.watermark_buffer_strategy.tick();
         self.seal_current_epoch(new_epoch.curr)
             .instrument(tracing::info_span!("state_table_commit"))
-            .await
+            .await?;
+
+        // Refresh watermark cache if it is out of sync.
+        if USE_WATERMARK_CACHE && !self.watermark_cache.is_synced() {
+            if let Some(ref watermark) = self.prev_cleaned_watermark {
+                let range: (Bound<Once<Datum>>, Bound<Once<Datum>>) =
+                    (Included(once(Some(watermark.clone()))), Unbounded);
+                // NOTE(kwannoel): We buffer `pks` before inserting into watermark cache
+                // because we can't hold an immutable ref (via `iter_key_and_val_with_pk_range`)
+                // and a mutable ref (via `self.watermark_cache.insert`) at the same time.
+                // TODO(kwannoel): We can optimize it with:
+                // 1. Either use `RefCell`.
+                // 2. Or pass in a direct reference to LocalStateStore,
+                //    instead of referencing it indirectly from `self`.
+                //    Similar to how we do for pk_indices.
+                let mut pks = Vec::with_capacity(self.watermark_cache.capacity());
+                {
+                    let mut streams = vec![];
+                    for vnode in self.vnodes().iter_vnodes() {
+                        let stream = self
+                            .iter_key_and_val_with_pk_range(
+                                &range,
+                                vnode,
+                                PrefetchOptions::new_for_exhaust_iter(),
+                            )
+                            .await?;
+                        streams.push(Box::pin(stream));
+                    }
+                    let merged_stream = merge_sort(streams);
+                    pin_mut!(merged_stream);
+
+                    while pks.len() < self.watermark_cache.capacity() && let Some((pk, _row)) = merged_stream.next().await.transpose()? {
```

Comment:
Can we `take` here? Like https://github.com/risingwavelabs/risingwave/blob/bac048757258c9ab7f9a75aca4eb9c26be4ff8d1/src/stream/src/executor/sort_buffer.rs#L203-L225